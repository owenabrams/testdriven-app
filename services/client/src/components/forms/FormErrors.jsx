import React from 'react';
import { Box, List, ListItem, Typography } from '@mui/material';
import { CheckCircle, Cancel } from '@mui/icons-material';

const FormErrors = (props) => {
  return (
    <Box sx={{ mb: 2 }}>
      <List className="validation-list" sx={{ p: 0 }}>
        {props.formRules.map((rule) => (
          <ListItem 
            key={rule.id}
            className={rule.valid ? "success" : "error"}
            sx={{ 
              py: 0.5,
              px: 0,
              display: 'flex',
              alignItems: 'center',
              color: rule.valid ? '#23D160' : '#FF3860'
            }}
          >
            {rule.valid ? (
              <CheckCircle sx={{ mr: 1, fontSize: 16, color: '#23D160' }} />
            ) : (
              <Cancel sx={{ mr: 1, fontSize: 16, color: '#FF3860' }} />
            )}
            <Typography 
              variant="body2" 
              sx={{ 
                color: rule.valid ? '#23D160' : '#FF3860',
                fontSize: '0.875rem'
              }}
            >
              {rule.name}
            </Typography>
          </ListItem>
        ))}
      </List>
    </Box>
  );
};

export default FormErrors;