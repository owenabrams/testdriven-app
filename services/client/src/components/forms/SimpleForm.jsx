import React, { useState } from 'react';
import { Navigate } from 'react-router-dom';
import axios from 'axios';
import {
  Box,
  Typography,
  TextField,
  Button,
  Paper,
  Container,
  Divider
} from '@mui/material';
import {
  PersonAdd as PersonAddIcon,
  Login as LoginIcon
} from '@mui/icons-material';

const SimpleForm = (props) => {
  const [formData, setFormData] = useState({
    username: '',
    email: '',
    password: ''
  });

  const handleChange = (event) => {
    const { name, value } = event.target;
    setFormData(prev => ({
      ...prev,
      [name]: value
    }));
  };

  const handleSubmit = async (event) => {
    event.preventDefault();
    const formType = props.formType;
    
    const data = {
      email: formData.email,
      password: formData.password
    };
    
    if (formType === 'Register') {
      data.username = formData.username;
    }
    
    const url = `${process.env.REACT_APP_USERS_SERVICE_URL}/auth/${formType.toLowerCase()}`;
    
    try {
      const res = await axios.post(url, data);
      setFormData({ username: '', email: '', password: '' });
      // Create success message before authentication to ensure it appears
      props.createMessage('Successfully ' + (formType === 'Login' ? 'logged in' : 'registered') + '!', 'success');
      await props.loginUser(res.data.auth_token);
    } catch (err) {
      console.log(err);
      if (formType === 'Login') {
        props.createMessage('Login failed. Please check your credentials.', 'error');
      }
      if (formType === 'Register') {
        props.createMessage('Registration failed. User may already exist.', 'error');
      }
    }
  };

  if (props.isAuthenticated) {
    return <Navigate to="/" />;
  }

  const isLogin = props.formType === 'Login';
  const isRegister = props.formType === 'Register';

  return (
    <Container maxWidth="sm">
      <Box sx={{ mt: 4, mb: 4 }}>
        <Paper elevation={3} sx={{ p: 4, borderRadius: 2 }}>
          <Box sx={{ display: 'flex', alignItems: 'center', mb: 3 }}>
            {isLogin && <LoginIcon sx={{ mr: 2, fontSize: 32, color: 'primary.main' }} />}
            {isRegister && <PersonAddIcon sx={{ mr: 2, fontSize: 32, color: 'primary.main' }} />}
            <Typography variant="h4" component="h1" sx={{ fontWeight: 'bold' }}>
              {isLogin ? 'Login' : 'Register'}
            </Typography>
          </Box>
          
          <Divider sx={{ mb: 3 }} />
          
          <Box component="form" onSubmit={handleSubmit}>
            {isRegister && (
              <TextField
                name="username"
                label="Username"
                type="text"
                placeholder="Enter a username"
                required
                fullWidth
                margin="normal"
                value={formData.username}
                onChange={handleChange}
                variant="outlined"
                sx={{ mb: 2 }}
              />
            )}
            
            <TextField
              name="email"
              label="Email Address"
              type="email"
              placeholder="Enter an email address"
              required
              fullWidth
              margin="normal"
              value={formData.email}
              onChange={handleChange}
              variant="outlined"
              sx={{ mb: 2 }}
            />
            
            <TextField
              name="password"
              label="Password"
              type="password"
              placeholder="Enter a password"
              required
              fullWidth
              margin="normal"
              value={formData.password}
              onChange={handleChange}
              variant="outlined"
              sx={{ mb: 3 }}
            />
            
            <Button
              type="submit"
              variant="contained"
              fullWidth
              size="large"
              sx={{ py: 1.5, fontSize: '1.1rem', fontWeight: 'bold' }}
            >
              {isLogin ? 'Login' : 'Register'}
            </Button>
          </Box>
        </Paper>
      </Box>
    </Container>
  );
};

export default SimpleForm;