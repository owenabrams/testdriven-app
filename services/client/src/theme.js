import { createTheme } from '@mui/material/styles';

// Create a Material-UI theme inspired by TestDriven.io
const theme = createTheme({
  palette: {
    primary: {
      main: '#1976d2', // Beautiful blue
      dark: '#115293',
      light: '#42a5f5',
      contrastText: '#ffffff',
    },
    secondary: {
      main: '#ff4081', // Pink accent
      dark: '#c60055',
      light: '#ff79b0',
      contrastText: '#ffffff',
    },
    background: {
      default: '#f5f5f5',
      paper: '#ffffff',
    },
    text: {
      primary: '#333333',
      secondary: '#666666',
    },
  },
  typography: {
    fontFamily: '"Roboto", "Helvetica", "Arial", sans-serif',
    h1: {
      fontWeight: 300,
      fontSize: '2.5rem',
    },
    h2: {
      fontWeight: 300,
      fontSize: '2rem',
    },
    h3: {
      fontWeight: 400,
      fontSize: '1.75rem',
    },
    h4: {
      fontWeight: 400,
      fontSize: '1.5rem',
    },
    h5: {
      fontWeight: 400,
      fontSize: '1.25rem',
    },
    h6: {
      fontWeight: 500,
      fontSize: '1rem',
    },
    button: {
      textTransform: 'none', // Don't uppercase buttons
      fontWeight: 500,
    },
  },
  shape: {
    borderRadius: 8, // Rounded corners
  },
  components: {
    MuiButton: {
      styleOverrides: {
        root: {
          borderRadius: 8,
          padding: '8px 16px',
          boxShadow: 'none',
          '&:hover': {
            boxShadow: '0 2px 4px rgba(0,0,0,0.1)',
          },
        },
        contained: {
          boxShadow: '0 2px 4px rgba(0,0,0,0.1)',
          '&:hover': {
            boxShadow: '0 4px 8px rgba(0,0,0,0.15)',
          },
        },
      },
    },
    MuiAppBar: {
      styleOverrides: {
        root: {
          boxShadow: '0 2px 4px rgba(0,0,0,0.1)',
        },
      },
    },
    MuiCard: {
      styleOverrides: {
        root: {
          boxShadow: '0 2px 8px rgba(0,0,0,0.1)',
          borderRadius: 12,
        },
      },
    },
    MuiTextField: {
      styleOverrides: {
        root: {
          '& .MuiOutlinedInput-root': {
            borderRadius: 8,
          },
        },
      },
    },
  },
});

export default theme;
