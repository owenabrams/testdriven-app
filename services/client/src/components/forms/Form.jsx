import React, { useState, useEffect, useCallback } from 'react';
import { Navigate } from 'react-router-dom';
import axios from 'axios';
import {
  Box,
  Typography,
  TextField,
  Paper,
  Container,
  Divider
} from '@mui/material';
import {
  PersonAdd as PersonAddIcon,
  Login as LoginIcon
} from '@mui/icons-material';
import FormErrors from './FormErrors';
import { registerFormRules, loginFormRules } from './form-rules';

const Form = (props) => {
  // Local state management for form data
  const [formData, setFormData] = useState({
    username: '',
    email: '',
    password: ''
  });

  // Form validation state
  const [registerRules, setRegisterRules] = useState(registerFormRules);
  const [loginRules, setLoginRules] = useState(loginFormRules);
  const [valid, setValid] = useState(false);

  // Clear form function
  const clearForm = useCallback(() => {
    setFormData({
      username: '',
      email: '',
      password: ''
    });
  }, []);

  // Email validation
  const validateEmail = useCallback((email) => {
    // eslint-disable-next-line
    var re = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
    return re.test(email);
  }, []);

  // Reset all validation rules
  const resetRules = () => {
    const resetRegisterRules = registerFormRules.map(rule => ({ ...rule, valid: false }));
    const resetLoginRules = loginFormRules.map(rule => ({ ...rule, valid: false }));
    setRegisterRules(resetRegisterRules);
    setLoginRules(resetLoginRules);
    setValid(false);
  };

  // Check if all rules are valid
  const allTrue = useCallback(() => {
    const formRules = props.formType === 'Register' ? registerRules : loginRules;
    return formRules.every(rule => rule.valid);
  }, [props.formType, registerRules, loginRules]);

  // Validate form
  const validateForm = useCallback(() => {
    // Validate register form
    if (props.formType === 'Register') {
      const updatedRules = [...registerFormRules];
      if (formData.username.length > 5) updatedRules[0].valid = true;
      if (formData.email.length > 5) updatedRules[1].valid = true;
      if (validateEmail(formData.email)) updatedRules[2].valid = true;
      if (formData.password.length > 10) updatedRules[3].valid = true;
      setRegisterRules(updatedRules);
      
      setValid(updatedRules.every(rule => rule.valid));
    }
    
    // Validate login form
    if (props.formType === 'Login') {
      const updatedRules = [...loginFormRules];
      if (formData.email.length > 0) updatedRules[0].valid = true;
      if (formData.password.length > 0) updatedRules[1].valid = true;
      setLoginRules(updatedRules);
      
      setValid(updatedRules.every(rule => rule.valid));
    }
  }, [formData, props.formType, validateEmail]);

  // Handle form input changes
  const handleFormChange = useCallback((event) => {
    const { name, value } = event.target;
    setFormData(prev => ({
      ...prev,
      [name]: value
    }));
  }, []);

  // Handle form submission
  const handleUserFormSubmit = useCallback(async (event) => {
    event.preventDefault();
    const formType = props.formType;
    
    const data = {
      email: formData.email,
      password: formData.password
    };
    
    if (formType === 'Register') {
      data.username = formData.username;
    }
    
    const url = `${process.env.REACT_APP_USERS_SERVICE_URL}/auth/${formType.toLowerCase()}`;
    
    try {
      const res = await axios.post(url, data);
      clearForm();
      props.loginUser(res.data.auth_token);
    } catch (err) {
      console.log(err);
      if (formType === 'Login') {
        props.createMessage('User does not exist.', 'danger');
      }
      if (formType === 'Register') {
        props.createMessage('That user already exists.', 'danger');
      }
    }
  }, [formData, props.formType, props.loginUser, props.createMessage, clearForm]);

  // Clear form on mount
  useEffect(() => {
    clearForm();
    resetRules();
  }, [clearForm]);

  // Clear form when formType changes (modern equivalent of componentWillReceiveProps)
  useEffect(() => {
    clearForm();
    resetRules();
  }, [props.formType, clearForm]);

  // Validate form when data changes
  useEffect(() => {
    validateForm();
  }, [validateForm]);

  if (props.isAuthenticated) {
    return <Navigate to="/" />;
  }

  const isLogin = props.formType === 'Login';
  const isRegister = props.formType === 'Register';
  const currentFormRules = isRegister ? registerRules : loginRules;

  return (
    <Container maxWidth="sm">
      <Box sx={{ mt: 4, mb: 4 }}>
        <Paper elevation={3} sx={{ p: 4, borderRadius: 2 }}>
          <Box sx={{ display: 'flex', alignItems: 'center', mb: 3 }}>
            {isLogin && <LoginIcon sx={{ mr: 2, fontSize: 32, color: 'primary.main' }} />}
            {isRegister && <PersonAddIcon sx={{ mr: 2, fontSize: 32, color: 'primary.main' }} />}
            <Typography variant="h4" component="h1" sx={{ fontWeight: 'bold' }}>
              {isLogin ? 'Log In' : 'Register'}
            </Typography>
          </Box>
          
          <Divider sx={{ mb: 3 }} />
          
          <FormErrors
            formType={props.formType}
            formRules={currentFormRules}
          />
          
          <Box component="form" onSubmit={handleUserFormSubmit}>
            {isRegister && (
              <TextField
                name="username"
                label="Username"
                type="text"
                placeholder="Enter a username"
                required
                fullWidth
                margin="normal"
                value={formData.username}
                onChange={handleFormChange}
                variant="outlined"
                sx={{ mb: 2 }}
              />
            )}
            
            <TextField
              name="email"
              label="Email Address"
              type="email"
              placeholder="Enter an email address"
              required
              fullWidth
              margin="normal"
              value={formData.email}
              onChange={handleFormChange}
              variant="outlined"
              sx={{ mb: 2 }}
            />
            
            <TextField
              name="password"
              label="Password"
              type="password"
              placeholder="Enter a password"
              required
              fullWidth
              margin="normal"
              value={formData.password}
              onChange={handleFormChange}
              variant="outlined"
              sx={{ mb: 3 }}
            />
            
            <button
              type="submit"
              disabled={!valid}
              style={{
                width: '100%',
                padding: '12px',
                fontSize: '1.1rem',
                fontWeight: 'bold',
                backgroundColor: valid ? '#1976d2' : '#ccc',
                color: 'white',
                border: 'none',
                borderRadius: '4px',
                cursor: valid ? 'pointer' : 'not-allowed',
                opacity: valid ? 1 : 0.6
              }}
            >
              Submit
            </button>
          </Box>
        </Paper>
      </Box>
    </Container>
  );
};

export default Form;
