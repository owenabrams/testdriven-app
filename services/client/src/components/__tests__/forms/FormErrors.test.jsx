import React from 'react';
import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { ThemeProvider } from '@mui/material/styles';
import { CssBaseline } from '@mui/material';
import FormErrors from '../../forms/FormErrors';
import { registerFormRules, loginFormRules } from '../../forms/form-rules';
import theme from '../../../theme';

// Test wrapper component with all necessary providers
const TestWrapper = ({ children }) => (
  <ThemeProvider theme={theme}>
    <CssBaseline />
    <BrowserRouter>
      {children}
    </BrowserRouter>
  </ThemeProvider>
);

const testData = [
  {
    formType: 'Register',
    formRules: registerFormRules,
    expectedRules: [
      'Username must be greater than 5 characters.',
      'Email must be greater than 5 characters.',
      'Email must be a valid email address.',
      'Password must be greater than 10 characters.'
    ]
  },
  {
    formType: 'Login',
    formRules: loginFormRules,
    expectedRules: [
      'Email is required.',
      'Password is required.'
    ]
  }
];

describe('FormErrors Component', () => {
  testData.forEach((el) => {
    test(`FormErrors (with ${el.formType} form) renders properly`, () => {
      const { container } = render(
        <TestWrapper>
          <FormErrors 
            formType={el.formType}
            formRules={el.formRules}
          />
        </TestWrapper>
      );
      
      // Check that all expected rules are rendered
      el.expectedRules.forEach(rule => {
        expect(container.textContent).toContain(rule);
      });
      
      // Check that the correct number of list items are rendered
      const listItems = container.querySelectorAll('li');
      expect(listItems.length).toBe(el.expectedRules.length);
    });

    test(`FormErrors (with ${el.formType} form) renders a snapshot properly`, () => {
      const { container } = render(
        <TestWrapper>
          <FormErrors 
            formType={el.formType}
            formRules={el.formRules}
          />
        </TestWrapper>
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});