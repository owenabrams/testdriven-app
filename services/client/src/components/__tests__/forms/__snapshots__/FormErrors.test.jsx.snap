// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormErrors Component FormErrors (with Login form) renders a snapshot properly 1`] = `
<div
  class="MuiBox-root css-1qm1lh"
>
  <ul
    class="MuiList-root MuiList-padding validation-list css-duwaa2-MuiList-root"
  >
    <li
      class="MuiListItem-root MuiListItem-gutters MuiListItem-padding error css-vaqbqu-MuiListItem-root"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-197gm3n-MuiSvgIcon-root"
        data-testid="CancelIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
        />
      </svg>
      <p
        class="MuiTypography-root MuiTypography-body2 css-nhkaa8-MuiTypography-root"
      >
        Email is required.
      </p>
    </li>
    <li
      class="MuiListItem-root MuiListItem-gutters MuiListItem-padding error css-vaqbqu-MuiListItem-root"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-197gm3n-MuiSvgIcon-root"
        data-testid="CancelIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
        />
      </svg>
      <p
        class="MuiTypography-root MuiTypography-body2 css-nhkaa8-MuiTypography-root"
      >
        Password is required.
      </p>
    </li>
  </ul>
</div>
`;

exports[`FormErrors Component FormErrors (with Register form) renders a snapshot properly 1`] = `
<div
  class="MuiBox-root css-1qm1lh"
>
  <ul
    class="MuiList-root MuiList-padding validation-list css-duwaa2-MuiList-root"
  >
    <li
      class="MuiListItem-root MuiListItem-gutters MuiListItem-padding error css-vaqbqu-MuiListItem-root"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-197gm3n-MuiSvgIcon-root"
        data-testid="CancelIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
        />
      </svg>
      <p
        class="MuiTypography-root MuiTypography-body2 css-nhkaa8-MuiTypography-root"
      >
        Username must be greater than 5 characters.
      </p>
    </li>
    <li
      class="MuiListItem-root MuiListItem-gutters MuiListItem-padding error css-vaqbqu-MuiListItem-root"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-197gm3n-MuiSvgIcon-root"
        data-testid="CancelIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
        />
      </svg>
      <p
        class="MuiTypography-root MuiTypography-body2 css-nhkaa8-MuiTypography-root"
      >
        Email must be greater than 5 characters.
      </p>
    </li>
    <li
      class="MuiListItem-root MuiListItem-gutters MuiListItem-padding error css-vaqbqu-MuiListItem-root"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-197gm3n-MuiSvgIcon-root"
        data-testid="CancelIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
        />
      </svg>
      <p
        class="MuiTypography-root MuiTypography-body2 css-nhkaa8-MuiTypography-root"
      >
        Email must be a valid email address.
      </p>
    </li>
    <li
      class="MuiListItem-root MuiListItem-gutters MuiListItem-padding error css-vaqbqu-MuiListItem-root"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-197gm3n-MuiSvgIcon-root"
        data-testid="CancelIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
        />
      </svg>
      <p
        class="MuiTypography-root MuiTypography-body2 css-nhkaa8-MuiTypography-root"
      >
        Password must be greater than 10 characters.
      </p>
    </li>
  </ul>
</div>
`;
